src/Controller/Component/ModulesComponent.php

```diff
@@ -145,10 +145,39 @@ function (&$data, $key) use ($metaModules) {
             array_diff_key($metaModules, $modules),
             $pluginModules
         );
+        $this->modules = $this->modulesByRoleConfig($this->modules);
 
         return $this->modules;
     }
 
+    /**
+     * This filters modules and apply 'Role' config by user role, if any.
+     * Module can be "hidden": remove from $modules.
+     * Module can be "readonly": add "'readonly' => true" to module.
+     *
+     * @param array $modules The modules
+     * @return array
+     */
+    protected function modulesByRoleConfig(array $modules): array
+    {
+        $user = $this->getController()->Auth->user();
+        $role = (string)Hash::get($user, 'roles.0');
+        $hidden = (array)Configure::read(sprintf('Roles.%s.hidden', $role));
+        $readonly = (array)Configure::read(sprintf('Roles.%s.readonly', $role));
+        $keys = array_keys($modules);
+        foreach ($keys as $moduleName) {
+            if (in_array($moduleName, $hidden)) {
+                unset($modules[$moduleName]);
+                continue;
+            }
+            if (in_array($moduleName, $readonly)) {
+                $modules[$moduleName]['readonly'] = true;
```

Comment:
it's ok, but we could set `hints.allow` array to `array_diff($allow, ['POST', 'PATCH', 'DELETE'])` and don't change `PermsHelper::isAllowed()`

Reply:
in `modulesFromMeta`? not clear how to do that without changing something in `PermsHelper::isAllowed`.

Reply:
I suggest something like this

```php
// make sure $readonly contains valid module names
$readonly = array_intersect($readonly, array_keys($modules));
foreach ($readonly as $key) {
   $path = sprintf('%s.hints.allow', $key);
   $allow = (array)Hash::get($modules, $path);
   Hash::insert($modules, $path, array_diff($allow, ['POST', 'PATCH', 'DELETE']));
}
```
This way no change in `PermsHelper` is necessary since perms logic uses `hints.allow` only